src/shogun/multiclass/KNN.cpp

```diff
@@ -123,33 +120,47 @@ SGMatrix<index_t> KNN::nearest_neighbors()
 	distance->precompute_lhs();
 	distance->precompute_rhs();
 
-	//for each test example
-	for (auto i : SG_PROGRESS(range(n)))
+	switch (m_knn_solver)
 	{
-		COMPUTATION_CONTROLLERS
-		//lhs idx 0..num train examples-1 (i.e., all train examples) and rhs idx i
-		distances_lhs(dists,0,m_train_labels.vlen-1,i);
-
-		//fill in an array with 0..num train examples-1
-		for (int32_t j=0; j<m_train_labels.vlen; j++)
-			train_idxs[j]=j;
-
-		//sort the distance vector between test example i and all train examples
-		Math::qsort_index(dists.vector, train_idxs.vector, m_train_labels.vlen);
-
-#ifdef DEBUG_KNN
-		io::print("\nQuick sort query {}\n", i);
-		for (int32_t j=0; j<m_k; j++)
-			io::print("{} ", train_idxs[j]);
-		io::print("\n");
-#endif
-
-		//fill in the output the indices of the nearest neighbors
-		for (int32_t j=0; j<m_k; j++)
-			NN(j,i) = train_idxs[j];
+	case KNN_BRUTE:
+	{
+		//for each test example
+		for (auto i : SG_PROGRESS(range(n)))
+		{
+			COMPUTATION_CONTROLLERS
+			//lhs idx 0..num train examples-1 (i.e., all train examples) and rhs idx i
+			distances_lhs(dists,0,m_train_labels.vlen-1,i);
+
+			//fill in an array with 0..num train examples-1
+			train_idxs.range_fill(0);
+
+			//sort the distance vector between test example i and all train examples
+			Math::qsort_index(dists.vector, train_idxs.vector, m_train_labels.vlen);
+
+			SG_DEBUG("\nQuick sort query {}", i);
+			SG_DEBUG("{}", train_idxs.to_string());
+			SG_DEBUG("");
+
+			//fill in the output the indices of the nearest neighbors
+			sg_memcpy(&NN.matrix[m_k * i], train_idxs.vector, sizeof(index_t) * m_k);
```

Comment:
there are methods in sgmatrix/sgvector that do this I think? (make it more readable)

Reply:
Yes, that is true. There is a `set_column()` function that I can use.
However, for that, I will need to reduce the size of the `train_idxs` vector to`m_k` or,
copy just the first `m_k` elements to a new array before using the `set_column()` function.
Is that okay?

Reply:
(definitely no copying)

Reply:
Okay.